tapdb/rfq_policies.go

```diff
@@ -0,0 +1,533 @@
+		Scid:                uint64(accept.ShortChannelId()),
+		RfqID:               rfqIDArray(accept.ID),
+		Peer:                serializePeer(accept.Peer),
+		AssetID:             assetID,
+		AssetGroupKey:       groupKey,
+		RateCoefficient:     rateBytes,
+		RateScale:           accept.AssetRate.Rate.Scale,
+		ExpiryUnix:          uint64(expiry.Unix()),
+		MaxOutAssetAmt:      ptrUint64(accept.Request.AssetMaxAmt),
+		RequestAssetMaxAmt:  ptrUint64(accept.Request.AssetMaxAmt),
+		PriceOracleMetadata: accept.Request.PriceOracleMetadata,
+		RequestVersion:      ptrUint32(uint32(accept.Request.Version)),
+		AgreedAt:            time.Now().UTC(),
+	}
+
+	return s.storePolicy(ctx, record)
+}
+
+// StorePurchasePolicy persists a sell-accept policy.
+func (s *PersistedPolicyStore) StorePurchasePolicy(ctx context.Context,
+	acpt rfqmsg.SellAccept) error {
+
+	assetID, groupKey := specifierPointers(acpt.Request.AssetSpecifier)
+	rateBytes := coefficientBytes(acpt.AssetRate.Rate)
+	expiry := acpt.AssetRate.Expiry.UTC()
+	paymentMax := int64(acpt.Request.PaymentMaxAmt)
+
+	record := rfqPolicy{
+		PolicyType:            RfqPolicyTypeAssetPurchase,
+		Scid:                  uint64(acpt.ShortChannelId()),
+		RfqID:                 rfqIDArray(acpt.ID),
+		Peer:                  serializePeer(acpt.Peer),
+		AssetID:               assetID,
+		AssetGroupKey:         groupKey,
+		RateCoefficient:       rateBytes,
+		RateScale:             acpt.AssetRate.Rate.Scale,
+		ExpiryUnix:            uint64(expiry.Unix()),
+		PaymentMaxMsat:        ptrInt64(paymentMax),
+		RequestPaymentMaxMsat: ptrInt64(paymentMax),
+		PriceOracleMetadata:   acpt.Request.PriceOracleMetadata,
+		RequestVersion:        ptrUint32(uint32(acpt.Request.Version)),
+		AgreedAt:              time.Now().UTC(),
+	}
+
+	return s.storePolicy(ctx, record)
+}
+
+func (s *PersistedPolicyStore) storePolicy(ctx context.Context,
+	policy rfqPolicy) error {
+
+	writeOpts := WriteTxOption()
+	return s.db.ExecTx(ctx, writeOpts, func(q RfqPolicyStore) error {
+		_, err := q.InsertRfqPolicy(ctx, newInsertParams(policy))
+		if err != nil {
+			return fmt.Errorf("error inserting RFQ policy: %w", err)
+		}
+
+		return nil
+	})
+}
+
+// FetchAcceptedQuotes retrieves all non-expired policies from the database and
+// returns them as buy and sell accepts.
+func (s *PersistedPolicyStore) FetchAcceptedQuotes(ctx context.Context) (
+	[]rfqmsg.BuyAccept, []rfqmsg.SellAccept, error) {
+
+	readOpts := ReadTxOption()
+	var (
+		buyAccepts  []rfqmsg.BuyAccept
+		sellAccepts []rfqmsg.SellAccept
+	)
+	now := time.Now().UTC()
+
+	err := s.db.ExecTx(ctx, readOpts, func(q RfqPolicyStore) error {
+		rows, err := q.FetchActiveRfqPolicies(ctx)
+		if err != nil {
+			return fmt.Errorf("error fetching policies: %w", err)
+		}
+
+		for _, row := range rows {
+			policy := policyFromRow(row)
+
+			// Skip expired entries.
+			expiry := time.Unix(int64(policy.ExpiryUnix), 0).UTC()
+			if now.After(expiry) {
+				continue
+			}
```

Comment:
This sort of filtering makes more sense in the `FetchActiveRfqPolicies` query, since most entries will be expired normally.